File: deep_learning/logic_simulation/perceptron/perceptron.py
# A self-coded perceptron (single-neuron neural net)
import numpy as np
class Perceptron:
  def __init__(self, num_inputs=3, weights=[1,1,1]):
    # num_inputs is the number of inputs that will be included in a list of data
    self.num_inputs = num_inputs
    # weights are the weight assigned to each synapse for the neural network
    self.weights = weights

  def weighted_sum(self, inputs):
    """return the sum for all inputs multiplied by its weight"""
    weighted_sum = 0
    for i in range(self.num_inputs):
      weighted_sum += self.weights[i]*inputs[i]
    return weighted_sum

  def activation(self, z):
    """Return a result between 1 and 0 using a sigmoid function"""
    denominator = 1 + np.exp(-z)
    result = 1/denominator
    return result

  def threshold(self, value):
    """Return a classification threshold of either 0 or 1
       we use 0.5 as a threshold here for ease of understanding
    """
    threshold = 0.5
    if value >= threshold:
        return 1
    else:
        return 0
 
  def training(self, training_set):
    """Train the neural network"""
    # Start foundline with False until the neural network is able to 
    # Correctly come up with a line that can classify the dataset
    foundLine = False
    # Keep track of the total amount of iterations
    iterations = 0
    while not foundLine:
      print(f"Start Training... Iterations={iterations}")
      total_error = 0
      for inputs in training_set:
        # Start with random weights 1,1,1
        # And uses simple multiplication and addition to get a weighted sum
        # pass it through the activation function to get either a 1 or 0
        prediction_old = self.activation(self.weighted_sum(inputs))
        # pass the prediction through the threshold to be get a value of 0 or 1
        # for easy training
        prediction = self.threshold(prediction_old)
        print(f"[Predicted]{prediction}")
        actual = training_set[inputs]
        print(f"[Actual]{actual}")
        # Calculate the error by subtracting actual with predicted value
        # So error can either be 2 or -2
        # each piece of data in the dataset will have an error rate
        error = actual - prediction
        # total error is the sum of all errors in the entire dataset
        # Take the absolute value so that error can only be 2, so both
        # -2 and +2 will have to same effect to the neural network
        total_error += abs(error)
        for i in range(self.num_inputs):
          # modify the weight by using error multiply input, this will result in a
          # better weight that can work better with the dataset
          # This is back backpropagation
          self.weights[i] += error*inputs[i]
        print(f"[Weights]{self.weights}")
      # End the training process when the total error is 0
      if total_error == 0:
        foundLine = True
        print(f"Training completed! Total Error={total_error}")
      iterations += 1

  def predict(self, testing_set):
    """return the prediction and confidence for a given testing data"""
    # Run prediction by multiply inputs with the weight and map it
    # Through the activation function
    final_prob = 0
    probability = self.activation(self.weighted_sum(testing_set))
    prediction = self.threshold(probability)
    if prediction == 1:
        final_prob = probability
    else:
        final_prob = 1 - probability
    return [prediction, final_prob]

if __name__ == '__main__':
    cool_perceptron = Perceptron()
    # Use the prediction result of (A - (B + C) > 0) to build the dataset
    # So that we can actually check the model
    small_training_set = {(1,1,0):0, (2,0,1):1, (3,2,0):1, (4,5,1):0}
    cool_perceptron.training(small_training_set)
    
    # Testing the model
    testing_set = (3,1,1)
    print(f"Start Prediction, testing_set={testing_set}")
    expected = 1
    result = cool_perceptron.predict(testing_set)
    print(f"result:{result[0]}\tprobability:{result[1]}")
    print(f"expected:{expected}")
    print(f"error={result[0]-expected}")

File: deep_learning/logic_simulation/perceptron_cc/perceptron.c
//
// Copyright © 2021 Max Xie, All rights reserved.
//
#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>
#include <math.h>
#include <stdlib.h>

#define DEFAULT_NUM_INPUTS 3
#define SAMPLE_SIZE 4
#define NUM_INPUTS 3

struct Perceptron {
    uint32_t num_inputs;
    double *weights;
};

typedef struct Perceptron Perceptron;

// Create a Perceptron
Perceptron *pc_create(uint32_t num_inputs, double initial_weight) {
    Perceptron *pc = (Perceptron *)malloc(sizeof(Perceptron));
    if (pc) {
        pc->num_inputs = (num_inputs == 0) ? DEFAULT_NUM_INPUTS : num_inputs;
        uint32_t weight = initial_weight;
        pc->weights = (double *)calloc(num_inputs, sizeof(double));
        // Initialize the weights
        for (uint32_t i = 0; i < pc->num_inputs; i++) {
            pc->weights[i] = weight;
        }
    }
    return pc;
}

// Delete a Perceptron
void pc_delete(Perceptron **pc) {
    if (*pc && (*pc)->weights) {
        free((*pc)->weights);
        free(*pc);
    }
    *pc = NULL;
}

// Calculate the weighted sum and return
double weighted_sum(Perceptron *pc, uint32_t *inputs) {
    double weighted_sum = 0.0;
    for (uint32_t i = 0; i < pc->num_inputs; i++) {
        weighted_sum += pc->weights[i] * inputs[i];
    }
    return weighted_sum;
}

// Return a result between 1 and 0 using sigmoid function
double activation(int32_t z) {
    double denominator = 1.0 + exp(-z);
    double result = 1.0/denominator;
    return result;
}

// Return a classification threshold of either 0 or 1
// we use 0.5 as a threshold here for ease of understanding
uint32_t threshold(double value) {
    double threshold = 0.5;
    if (value >= threshold) {
        return 1;
    }
    return 0;
}

// Train the neural network
void training(Perceptron *pc, uint32_t training_set[SAMPLE_SIZE][NUM_INPUTS], uint32_t *expected_data, uint32_t training_data_size) {
    bool found_line = false;
    uint32_t iterations = 0;
    while (!found_line) {
        printf("iterations=%d\n", iterations);
        double total_error = 0;
        for (uint32_t i = 0; i < training_data_size; i++) {
            // Calculate the weighted_sum and pass it through activation function
            double prediction_old = activation(weighted_sum(pc, training_set[i]));
            uint32_t prediction = threshold(prediction_old);
            // Check prediction against predicted data, and calculate the loss
            uint32_t actual = expected_data[i];
            int32_t error = actual - prediction;
            total_error += abs(error);
            // adjust the weight
            for (uint32_t j = 0; j < pc->num_inputs; j++) {
                pc->weights[j] += (int32_t)(error * training_set[i][j]);
            }
        }
        printf("[error] %f\n", total_error);
        // End the training process when the total error is 0
        if (total_error < 0.01) {
            found_line = true;
            printf("Training completed! Total Error=%.5f\n", total_error);
        }
        iterations++;
    }
}

// Return the prediction and confidence for a given testing data set
void predict(Perceptron *pc, uint32_t *testing_set) {
    double final_prob = 0.0;
    // Calculate the weighted_sum and pass it through activation function
    double probability = activation(weighted_sum(pc, testing_set));
    uint32_t prediction = threshold(probability);

    if (prediction == 1) {
        final_prob = probability;
    } else {
        final_prob = 1 - probability;
    }
    printf("Prediction: %d\n", prediction);
    printf("probability: %.5f\n", final_prob);
}

int main() {
    // Create a Perceptron
    Perceptron *pc = pc_create(NUM_INPUTS, 1.0);
    // Create training data
    // Use the prediction result of (A - (B + C) > 0) to build the dataset
    uint32_t small_training_set[SAMPLE_SIZE][NUM_INPUTS] = {
        { 1, 1, 0 },
        { 2, 0, 1 },
        { 3, 2, 0 },
        { 4, 5, 1 }
    };
    // Create expected result
    uint32_t expected_result[SAMPLE_SIZE] = { 0, 1, 1, 0 };

    // Train the model
    training(pc, small_training_set, expected_result, SAMPLE_SIZE);

    // Test the model, true case
    uint32_t small_testing_data[NUM_INPUTS] = { 3, 1, 1 };
    printf("Start prediction, testing_set=");
    for (int i = 0; i < NUM_INPUTS; i++) {
        printf("%d ", small_testing_data[i]);
    }
    printf("\n");

    uint32_t expected = 1;
    predict(pc, small_testing_data);
    printf("Expected: %d\n", expected);

    // Test the model, false case
    uint32_t small_testing_data2[NUM_INPUTS] = { 3, 5, 6 };
    printf("Start prediction, testing_set=");
    for (int i = 0; i < NUM_INPUTS; i++) {
        printf("%d ", small_testing_data2[i]);
    }
    printf("\n");

    uint32_t expected2 = 0;
    predict(pc, small_testing_data2);
    printf("Expected: %d\n", expected2);

    // garbage collection
    pc_delete(&pc);
}
